# Table of contents

0.1 Importing Libraries

0.2 Importing Data

0.3 Exploring Original Dfs

0.4 Merge

0.5 Checking Merge

0.6 Removing Merge Flag

0.7 Exporting Merge

### 0.1 Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

### 0.2 Importing Data

In [2]:
# Identify the file pathway to data files
path = r'C:\Users\CJ\Documents\_CJ-Stuff\Career Foundry\Data Immersion\Ach 4 - Python\2023-03 Instacart Basket Analysis'

In [3]:
# Import data from 'customers_clean.pkl'
cust = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'customers_clean.pkl'))

In [4]:
# Import data from 'orders_products_merged_4.pkl'
orders = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged2.pkl'))

### 0.3 Exploring Original Dfs

In [5]:
cust.shape

(206209, 8)

In [6]:
orders.shape

(32434489, 26)

The merged df should have 32,434,489 rows and 34 columns
(8 + 26 + 1 merge flag - 1 key overlap)

In [7]:
cust.head()

user_id  gender       state  age date_joined  n_dependants fam_status  \
0    26711  Female    Missouri   48  2017-01-01             3    married   
1    33890  Female  New Mexico   36  2017-01-01             0     single   
2    65803    Male       Idaho   35  2017-01-01             2    married   
3   125935  Female        Iowa   40  2017-01-01             0     single   
4   130797  Female    Maryland   26  2017-01-01             1    married   

   income  
0  165665  
1   59285  
2   99568  
3   42049  
4   40374

In [8]:
orders.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  first_order_of_new_customer  product_id  \
0                     NaN                         True         196   
1                     NaN                         True       14084   
2                     NaN                         True       12427   
3                     NaN                         True       26088   
4                     NaN                         True       26405   

   add_to_cart_order  reordered  ...        price_range     busiest_day  \
0                  1          0  ...  Mid-range product  Regularly busy   
1                  2          0  ...  Mid-range product  Regularly busy   
2                  3          0  ...  Low-range product  Regularly busy   
3                  4          0  ...  Low-range product  Regularly busy   
4                  5          0  ...  Low-range product  Regularly busy   

          busiest_days  busiest_period_of_day  max_order  loyalty_flag  \
0  Regularly busy days         Average orders         10  New customer   
1  Regularly busy days         Average orders         10  New customer   
2  Regularly busy days         Average orders         10  New customer   
3  Regularly busy days         Average orders         10  New customer   
4  Regularly busy days         Average orders         10  New customer   

  mean_spending spender_type median_days_since_ordering  frequent_shopper  
0      6.367797  Low spender                       20.0  Regular customer  
1      6.367797  Low spender                       20.0  Regular customer  
2      6.367797  Low spender                       20.0  Regular customer  
3      6.367797  Low spender                       20.0  Regular customer  
4      6.367797  Low spender                       20.0  Regular customer  

[5 rows x 26 columns]

In [9]:
cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 8 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   user_id       206209 non-null  int64         
 1   gender        206209 non-null  category      
 2   state         206209 non-null  category      
 3   age           206209 non-null  int16         
 4   date_joined   206209 non-null  datetime64[ns]
 5   n_dependants  206209 non-null  int8          
 6   fam_status    206209 non-null  category      
 7   income        206209 non-null  int64         
dtypes: category(3), datetime64[ns](1), int16(1), int64(2), int8(1)
memory usage: 5.9 MB


In [10]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32434489 entries, 0 to 32434488
Data columns (total 26 columns):
 #   Column                       Dtype   
---  ------                       -----   
 0   order_id                     int64   
 1   user_id                      int64   
 2   order_number                 int16   
 3   order_day_of_week            int8    
 4   order_hour_of_day            int8    
 5   days_since_prior_order       float16 
 6   first_order_of_new_customer  bool    
 7   product_id                   int32   
 8   add_to_cart_order            int16   
 9   reordered                    int8    
 10  product_name                 object  
 11  aisle_id                     float16 
 12  department_id                float16 
 13  price                        float32 
 14  product_info_available       bool    
 15  price_available              bool    
 16  price_range                  category
 17  busiest_day                  category
 18  busiest_days        

The merge should be on user_id

In [11]:
cust.isnull().sum()

user_id         0
gender          0
state           0
age             0
date_joined     0
n_dependants    0
fam_status      0
income          0
dtype: int64

In [12]:
orders.isnull().sum()

order_id                             0
user_id                              0
order_number                         0
order_day_of_week                    0
order_hour_of_day                    0
days_since_prior_order         2078068
first_order_of_new_customer          0
product_id                           0
add_to_cart_order                    0
reordered                            0
product_name                     30770
aisle_id                         30770
department_id                    30770
price                            35897
product_info_available               0
price_available                      0
price_range                      35897
busiest_day                          0
busiest_days                         0
busiest_period_of_day                0
max_order                            0
loyalty_flag                         0
mean_spending                        0
spender_type                         0
median_days_since_ordering           0
frequent_shopper         

In [13]:
orders['first_order_of_new_customer'].value_counts()

False    30356421
True      2078068
Name: first_order_of_new_customer, dtype: int64

In [14]:
orders['product_info_available'].value_counts()

True     32403719
False       30770
Name: product_info_available, dtype: int64

In [15]:
orders['price_available'].value_counts()

True     32398592
False       35897
Name: price_available, dtype: int64

Each of the fields with null/NaN values have appropriate, corresponding flags.

### 0.4 Merge

In [16]:
# An exploratory outer merge confirmed that an inner merge 
# could also be used to capture all records.
order_products_all = orders.merge(cust, on = ['user_id'], how = 'inner', indicator= True)

### 0.5 Checking Merge

In [17]:
#Checking for whether there are any rows that don't have a match
order_products_all['_merge'].value_counts()

both          32434489
left_only            0
right_only           0
Name: _merge, dtype: int64

In [18]:
order_products_all.shape

(32434489, 34)

Since the number of rows matches the rows in the orders df
AND the total number of columns =
(columns in orders df + the columns in customers + merge flag - 1 key column)
this merge can be considered successful.

In [19]:
order_products_all.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  first_order_of_new_customer  product_id  \
0                     NaN                         True         196   
1                     NaN                         True       14084   
2                     NaN                         True       12427   
3                     NaN                         True       26088   
4                     NaN                         True       26405   

   add_to_cart_order  reordered  ... median_days_since_ordering  \
0                  1          0  ...                       20.0   
1                  2          0  ...                       20.0   
2                  3          0  ...                       20.0   
3                  4          0  ...                       20.0   
4                  5          0  ...                       20.0   

   frequent_shopper  gender    state  age  date_joined n_dependants  \
0  Regular customer  Female  Alabama   31   2019-02-17            3   
1  Regular customer  Female  Alabama   31   2019-02-17            3   
2  Regular customer  Female  Alabama   31   2019-02-17            3   
3  Regular customer  Female  Alabama   31   2019-02-17            3   
4  Regular customer  Female  Alabama   31   2019-02-17            3   

  fam_status income _merge  
0    married  40423   both  
1    married  40423   both  
2    married  40423   both  
3    married  40423   both  
4    married  40423   both  

[5 rows x 34 columns]

In [20]:
# Checking for nulls across df
order_products_all.isnull().sum()

order_id                             0
user_id                              0
order_number                         0
order_day_of_week                    0
order_hour_of_day                    0
days_since_prior_order         2078068
first_order_of_new_customer          0
product_id                           0
add_to_cart_order                    0
reordered                            0
product_name                     30770
aisle_id                         30770
department_id                    30770
price                            35897
product_info_available               0
price_available                      0
price_range                      35897
busiest_day                          0
busiest_days                         0
busiest_period_of_day                0
max_order                            0
loyalty_flag                         0
mean_spending                        0
spender_type                         0
median_days_since_ordering           0
frequent_shopper         

The null/NaN values are the same as before the merge.

### 0.6 Removing the merge flag so it does not intefere with future merges

In [21]:
order_products_all = order_products_all.drop(columns = ['_merge'])

In [22]:
# Confirming that only the merge column was dropped.
order_products_all.shape

(32434489, 33)

In [23]:
order_products_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32434489 entries, 0 to 32434488
Data columns (total 33 columns):
 #   Column                       Dtype         
---  ------                       -----         
 0   order_id                     int64         
 1   user_id                      int64         
 2   order_number                 int16         
 3   order_day_of_week            int8          
 4   order_hour_of_day            int8          
 5   days_since_prior_order       float16       
 6   first_order_of_new_customer  bool          
 7   product_id                   int32         
 8   add_to_cart_order            int16         
 9   reordered                    int8          
 10  product_name                 object        
 11  aisle_id                     float16       
 12  department_id                float16       
 13  price                        float32       
 14  product_info_available       bool          
 15  price_available              bool          
 16

### 0.7 Exporting the final df

In [24]:
order_products_all.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_all.pkl'))